### PR TITLE
PT-2037: option --system-only for pt-stalk

### PR DESCRIPTION
**Pull request by svetasmirnova**
Fix for PT-2037: Add option --skip-mysql or --system-only for pt-stalk

### CODE REVIEW

File: bin/pt-stalk [hunk 1]
```diff
@@ -796,84 +797,88 @@ collect() {
    local d="$1"  # directory to save results in
    local p="$2"  # prefix for each result file
 
-   local mysqld_pid=""
-   if [ ! "$OPT_MYSQL_ONLY" ]; then
-      mysqld_pid=$(_pidof mysqld | awk '{print $1; exit;}')
-   fi
+   if [ ! "$OPT_SYSTEM_ONLY" ]; then
```
Comment (denisok) [inline bin/pt-stalk:806]: instead of `if` maybe it is good idea to create new function for system collect? and just call it in the beginning and exit if "$OPT_SYSTEM_ONLY" is set. 

And do we need new key `OPT_SYSTEM_ONLY` at all? if there is no mysqld process - just don't execute mysqld section, if OPT_MYSQL_ONLY is set don't execute the system function

Comment (svetasmirnova) [reply bin/pt-stalk:806]: I made pt-stalk data collection more modular in commit d6e89940..623fdaec. Please review.

Regarding the option: while the main purpose for this patch is to allow pt-stalk to run on machines where MySQL is not installed, we may need to run it on machines where MySQL is installed, but not the source of the problem. For example, working with a customer who has a small MySQL server and large MongoDB or PostgreSQL that we are troubleshooting.

File: bin/pt-stalk [hunk 2]
```diff
@@ -128,7 +128,6 @@ OPT_ERRS=0        # How many command line option errors
 OPT_VERSION=""    # If --version was specified
 OPT_HELP=""       # If --help was specified
 OPT_ASK_PASS=""   # If --ask-pass was specified
-OPT_MYSQL_ONLY="" # If --mysql-only was specified
```
Comment (denisok) [inline bin/pt-stalk:131]: why did you remove this option ? you added another one so you should add it here. Probably it was missed from your previous commits.

File: bin/pt-stalk [hunk 3]
```diff
@@ -459,7 +458,6 @@ _parse_command_line() {
          else
             spec=$(grep "^short form:-$opt\$" "$PT_TMPDIR"/po/* | cut -d ':' -f 1)
             if [ -z "$spec"  ]; then
-               option_error "Unknown option: $real_opt"
```
Comment (denisok) [inline bin/pt-stalk:462]: why is this removed?

File: bin/pt-stalk [hunk 4]
```diff
@@ -796,9 +801,65 @@ collect() {
...
+         log "Removing empty file $file";
+         rm "$file"
+      fi
+   done
+}
+
+collect_mysql_data_one() {
    if [ ! "$OPT_MYSQL_ONLY" ]; then
-      mysqld_pid=$(_pidof mysqld | awk '{print $1; exit;}')
+      port=$($CMD_MYSQL $EXT_ARGV -ss -e 'SELECT @@port')
```
Comment (denisok) [inline bin/pt-stalk:861]: ok, is this some kind of improvement? Should we separate this to another ticket, so it could be more descriptive. Because adding it in the scope of this fix - it could be hard to figure out later why it was done.

File: bin/pt-stalk [hunk 5]
```diff
@@ -820,27 +881,27 @@ collect() {
...
       mysql_error_log="$(ls -l /proc/$mysqld_pid/fd | awk '/ 2 ->/{print $NF}')"
    fi
 
-   local tail_error_log_pid=""
    if [ "$mysql_error_log" -a ! "$OPT_MYSQL_ONLY" ]; then
       log "The MySQL error log seems to be $mysql_error_log"
       tail -f "$mysql_error_log" >"$d/$p-log_error" &
       tail_error_log_pid=$!
 
-      $CMD_MYSQLADMIN $EXT_ARGV debug
```
Comment (denisok) [inline bin/pt-stalk:836]: why we delete this?

File: bin/pt-stalk [hunk 6]
```diff
@@ -1171,11 +1219,11 @@ slave_status() {
       echo -e "\n$sql\n" >> $outfile
       $CMD_MYSQL $EXT_ARGV -e "$sql" >> $outfile
 
-      sql="SELECT * FROM replication_connection_status\G"
+      sql="SELECT * FROM performance_schema.replication_connection_status\G"
       echo -e "\n$sql\n" >> $outfile
       $CMD_MYSQL $EXT_ARGV -e "$sql" >> $outfile
 
-      sql="SELECT * FROM replication_applier_status JOIN replication_applier_status_by_coordinator USING(channel_name)\G"
+      sql="SELECT * FROM performance_schema.replication_applier_status JOIN performance_schema.replication_applier_status_by_coordinator USING(channel_name)\G"
```
Comment (denisok) [inline bin/pt-stalk:1226]: how did it work before ? or it will drop support for some old DB versions?

File: t/pt-stalk/pt-stalk.t [hunk 7]
```diff
@@ -542,20 +542,68 @@ $retval = system("$trunk/bin/pt-stalk --no-stalk --run-time 2 --sleep 2 --dest $
...
+$output = $output - `ls -l $dest | wc -l`;
 
 is(
    $output,
    0,
    "Retention test 5: retention by count works as expected"
 );
 
+# ###########################################################################
+# Test if option --system-only works correctly
```
Comment (denisok) [inline t/pt-stalk/pt-stalk.t:560]: what if both options are given ? could you add test for that and check how your system behaves ?

**Review by denisok — COMMENTED** [review]
Overall looks good, I have couple of questions.

File: bin/pt-stalk (hunk 5 above)
Comment (svetasmirnova) [reply bin/pt-stalk:836]: This is the fix for https://jira.percona.com/browse/PT-1340 that was removed by the fix for PT-1398

File: bin/pt-stalk (hunk 6 above)
Comment (svetasmirnova) [reply bin/pt-stalk:1226]: It did not work. The log contained error messages that table replication_applier_status is not found. I believe this was, again, the result of mix-up of library and tool code.

File: bin/pt-stalk (hunk 2 above)
Comment (svetasmirnova) [reply bin/pt-stalk:131]: Because this is part of the auto-copied code from lib/bash/parse_options.sh The tool works correctly without it

Comment (svetasmirnova) [reply bin/pt-stalk:131]: Function _eval_po from parse_options.sh evaluates such options automatically and assigns value of mysql-only to OPT_MYSQL_ONLY, value system-only to OPT_SYSTEM_ONLY, and so on

File: bin/pt-stalk (hunk 3 above)
Comment (svetasmirnova) [reply bin/pt-stalk:462]: This is fix for https://jira.percona.com/browse/PT-1720 that was not added to pt-stalk, because modules were not updated properly

File: bin/pt-stalk (hunk 4 above)
Comment (svetasmirnova) [reply bin/pt-stalk:861]: This is the fix for https://jira.percona.com/browse/PT-1398

File: t/pt-stalk/pt-stalk.t (hunk 7 above)
Comment (svetasmirnova) [reply t/pt-stalk/pt-stalk.t:560]: Test case added. See commit 3767e57e151fe7e3f9631d5c736334c1b8fc431b

**Review by denisok — APPROVED** [review]
